### PR TITLE
JavaScript: Replace all uses of tracked nodes with type tracking.

### PR DESCRIPTION
**Pull request by xiemaisi**
This has turned into a fairly big PR, with three layers of commits (which I'd be happy to submit as separate PRs if desired):

  - The first group of commits replaces all current uses of tracked nodes (in the socket.io and HTTP libraries) with type tracking.
  - The second group of commits tweaks caching structure and refactors a few predicates to avoid unhelpful magic exposed by the first group of commits. These are intended to be behaviour-preserving. 
  - The third group of commits extends type tracking with support for tracking through one level of properties, since otherwise we're losing results compared to tracked nodes. This entails a small change to the API, see the qldoc of `TypeTracker` and `TypeBackTracker` for details.

Performance is [pretty awesome](https://git.semmle.com/max/dist-compare-reports/blob/master/js/type-tracking-instead-of-tracked-nodes/report.md) (internal link). The new results on NodeGoat look like (seeded) true positives; the regexp injection on semmle-express-test is, I think, a false positive that is unrelated to this PR (we assume every call to `match` with the right arity does regexp matching); and the remote-property injection results are harmless (the object into which the properties are injected has a `null` prototype).

### CODE REVIEW

**Comment by xiemaisi** [comment]
> Apart from the flowsToSourceNode change I wouldn't mind adopting the same style

Yeah, `flowsToSourceNode` is horrible; I'll change to `ref`.

I found myself preferring `prev` and `next` over `t2` because it clarified in my mind which direction the tracking was happening in, and indeed I made a few mistakes that I caught that way. `s` is no good, though, I'll get rid of it.

I don't think there was any good reason for me changing the description of `t`, so I'll just revert to what you had earlier.

File: javascript/ql/src/semmle/javascript/Paths.qll [hunk 1]
```diff
@@ -236,3 +236,20 @@ private class ConcatPath extends PathExpr {
     result = this.(AddExpr).getAnOperand().(PathExpr).getSearchRoot(priority)
   }
 }
+
+/**
+ * An expression that appears in a syntactic position where it may represent a path.
+ *
+ * Examples include arguments to the CommonJS `require` function or AMD dependency arguments.
+ */
+abstract class PathExprCandidate extends Expr {
```
[the remark this answers is not in the document]
Comment (xiemaisi) [reply javascript/ql/src/semmle/javascript/Paths.qll]: Oh, good idea. It'll require rebuilding some snapshots, but seems worth doing.

**Comment by xiemaisi** [comment]
Naming suggestions addressed; running the import graph comparison now.

**Comment by xiemaisi** [comment]
No changes to `PathExpr::resolve` on our default projects. Is there anything else you'd like me to check?

**Comment by asger-semmle** [comment]
Good enough for me

**Pull request merged**